### PR TITLE
Adding a tool to create intervals for WGS variant calling.

### PR DESCRIPTION
**Pull request by nh13**

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+  private var counter: Int = 0
+  private var refName: String = ""
+  private var refIndex: Int = -1
+  private var start: Int = -1
+  private var end: Int = -1
+  private val intervals = ListBuffer[Interval]()
+
+  def add(locus: LocusInfo): Boolean = {
+    val coverage = locus.getRecordAndPositions.size()
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: Could be improved with https://github.com/samtools/htsjdk/pull/616/ since then we wouldn't have to create a new list each time.

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Creates a list of intervals for WGS variant calling.
+  */
+@clp(description =
+  """
+    |Creates a list of intervals for WGS variant calling.
+    |
+    |The output intervals will only contain sites that have both a minimum and maximum coverage as set by the user.
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: have both a -> between the

**Comment by tfenne** [comment]
A few quick thoughts:

1. By default SamLocusIterator filters out both secondary and supplementary alignments.  You might want to change that to include supplemental.
2. Do you want to do any filtering on mapq?  Maybe exclude mapq 0 or make that a parameter?
3. Would it be beneficial to also take in the reference and exclude bases that are non-ACGT in the reference?

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  if (maxToAccumulatePerLocus < 1.0f) throw new ValidationException("The scaling factor for maximum number of reads to accumulate per locus must be => 1.")
+
+  override def execute(): Unit = {
+    val progress: ProgressLogger = new ProgressLogger(logger, noun="positions", unit=10*1000*1000)
+    val in: SamReader = SamReaderFactory.make.open(input.toFile)
+    val locusIterator = new SamLocusIterator(in)
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: I think you probably want to `setIncludeIndels()` here too, otherwise you'll exclude any sites that are homozygously deleted.

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala (hunk 1 above)
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: True, although it's only a wrapping of an already-created list in an unmodifiable.

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    converter.intervalList.write(output.toFile)
+
+    CloserUtil.close(locusIterator)
+  }
+}
+
+private class CoveredLociToIntervals(val header: SAMFileHeader, val minCoverage: Int, val maxCoverage: Int) {
+
+  private var counter: Int = 0
+  private var refName: String = ""
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: Would be nicer to use `Option[]` here instead of `String`/`Int` with magic "no value" values.

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  private var refIndex: Int = -1
+  private var start: Int = -1
+  private var end: Int = -1
+  private val intervals = ListBuffer[Interval]()
+
+  def add(locus: LocusInfo): Boolean = {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: I find this whole method a little hard to follow, and I'm not convinced that it won't lose the last covered interval in the genome.  I would suggest making a four-way case statement along the lines of:

```scala
val acceptableCoverage = coverage < minCoverage || maxCoverage < coverage
val inCoverage = this.refIndex != -1

(inCoverage, acceptableCoverage) match {
    ...
}
```

Alternatively I'd suggest having test cases that actually test all the possible state transitions.

**Comment by tfenne** [comment]
@nh13 My only real concern here is handling all the state transitions in the `CoveredLociToIntervals` class.  It seems like it's handling a lot (good->not good, not good-> good, "skipped over" bases, switch to next reference, etc) implicitly.  If it were me I'd either want to handle more of those explicitly or have test cases for all of them.

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala (hunk 3 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: I'll need update the picard/htsjdk versions.  Will do.

Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: Waiting for https://github.com/samtools/htsjdk/pull/408 to be in master.

File: src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala (hunk 5 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: The last interval is checked in the `intervalList` method.  If we are the next base, we extend the interval, otherwise we start a new one.

Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/wgs/CreateWgsCallableIntervalList.scala]: @tfenne do we ever get a locus with unmapped reads?  If not, then I don't think `inCoverage` above is necessary.  I am bit lost about what is confusing here.
